### ISSUE TITLE
Support for Firestore Metadata fields (Id, CreateTimestamp, UpdateTimestamp, ReadTimestamp) in IFireStoreConverter<T>

### ISSUE THREAD

**Issue by ryandle** [googleapis/google-cloud-dotnet#5470]
**Is your feature request related to a problem? Please describe.**
I'm using a custom converter to convert my type to/from a document (IDictionary<string,object>) to avoid decorating my types with attributes. (I want my types to be shared in other libraries where I can't reference the Firestore lib.)

The custom converter works great, except I don't think it exposes the 4 metadata properties you can get using attribute serialization:

- FirestoreDocumentId
- FirestoreDocumentCreateTimestamp
- FirestoreDocumentUpdateTimestamp
- FirestoreDocumentReadTimestamp

 **Describe the solution you'd like**
It would be awesome of these 4 properties could be passed through the IFirestoreConverter<T> class

 **Describe alternatives you've considered**
I don't see any way to get these values using IFirestoreConverter<T> today.

 **Additional context**
I was exploring this scenario with the below types/converter:
```cs

using AutoMapper;
using Google.Cloud.Firestore;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace ConsoleApp1
{
    class Program
    {
        public static IMapper MapperInstance = new MapperConfiguration(cfg => { }).CreateMapper();

        static async Task Main(string[] args)
        {
            // Set GOOGLE_APPLICATION_CREDENTIALS env variable here

            FirestoreDbBuilder builder = new FirestoreDbBuilder();
            builder.ConverterRegistry = new ConverterRegistry();
            builder.ConverterRegistry.Add(new Converter());
            builder.ProjectId = "<project id>";

            FirestoreDb db = builder.Build();

            await db.Collection("Testing").AddAsync(new MyFireStoreType
            {
                Prop1 = "foo",
                Prop2 = 2
            });

            var result = await db.Collection("Testing").GetSnapshotAsync();

            MyFireStoreType deserializedType = result[0].ConvertTo<MyFireStoreType>();

            Console.WriteLine("Done");
        }
    }

    public sealed class MyFireStoreType : FireStoreBaseObject
    {
        public string Prop1 { get; set; }
        public int Prop2 { get; set; }
        public bool? Prop3 { get; set; }
    }

    public abstract class FireStoreBaseObject
    {
        public string FirestoreDocumentId { get; set; }

        public DateTimeOffset FirestoreDocumentCreateTimestamp { get; set; }

        public DateTimeOffset FirestoreDocumentUpdateTimestamp { get; set; }

        public DateTimeOffset FirestoreDocumentReadTimestamp { get; set; }
    }

    public class Converter : IFirestoreConverter<MyFireStoreType>
    {
        public MyFireStoreType FromFirestore(object value)
        {
            Dictionary<string, object> valueAsDict = (Dictionary<string, object>)value;
            // Issue - no Id, CreateTimestamp, UpdateTimestamp, ReadTimestamp values available in the above dictionary
            return Program.MapperInstance.Map<MyFireStoreType>(valueAsDict);
        }

        public object ToFirestore(MyFireStoreType value)
        {
            return Program.MapperInstance.Map<Dictionary<string, object>>(value);
        }
    }
}

```

**Comment by jskeet** [comment]
I'll try to have a look at what this would take tomorrow, but it's really important to note that the serialization aspects of this library have become pretty overwhelming - and significantly beyond what other Firestore libraries are doing, I believe. If there's a way of designing this in neatly, I'd be happy to do so - but I don't want to go to really extreme lengths.

(I do understand the desire to do this, just to be clear.)

As an aside, it would be useful if you could provide sample code written in a GA version of C# - I try to keep prereleases away from my work laptop.

**Comment by ryandle** [comment]
Thanks for giving this consideration @jskeet. I have updated my example above to not use pre-release C# 9 features.

**Comment by jskeet** [comment]
Okay, I've had an idea about how we should be able to make these available in an extensible, backwardly-compatible way. Basically you'd be able to configure additional properties that would be populated in the dictionary passed to `FromFirestore`. I'll work on a PR today/tomorrow with that in...

**Comment by jskeet** [comment]
If you could have a look at #5474 and see whether you think that would work for you, that would be really good to know. There are other alternatives we could explore, but I suspect they'd mostly involve a sort of "alternative/extended" IFirestoreConverter interface, which wouldn't be ideal... I thought that receiving the values within the existing dictionary is probably the cleanest option.

**Comment by ryandle** [comment]
Confirmed, that change will solve my scenario!

**Comment by jskeet** [comment]
This is now implemented in 2.3.0-beta01. (We used an attribute on the method instead of the type in the end.) If you could check that it actually works for you using that release, we can release it as 2.3.0 pretty soon.

**Comment by ryandle** [comment]
@jskeet I have pulled  2.3.0-beta01 and confirmed that release now supports my scenario! Thank you for your help.

**Comment by jskeet** [comment]
@ryandle: Thanks for the confirmation; that's really useful. Will plan for a 2.3.0 release soon.

**Comment by jskeet** [comment]
@ryandle: I released 2.3.0 yesterday - thanks for working so collaboratively on this.
